### PR TITLE
MM-41889 Gekidou Post Options menu - save/unsave option

### PR DESCRIPTION
**Pull request by avinashlng1080**
#### Summary
This PR adds the save/unsave functionality to the Post Options menu.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-41889

#### Release Note
```release-note
NONE
```
-->

### CODE REVIEW

File: app/actions/remote/preference.ts [hunk 1]
```diff
@@ -48,23 +51,56 @@ export const saveFavoriteChannel = async (serverUrl: string, channelId: string,
         return {error: `${serverUrl} database not found`};
     }
 
-    let client;
     try {
-        client = NetworkManager.getClient(serverUrl);
+        // Todo: @shaz I think you'll need to add the category handler here so that the channel is added/removed from the favorites category
```
Comment (larkox) [inline app/actions/remote/preference.ts:55]: Mention is nice, ticket is even better ;)

File: app/client/rest/preferences.ts [hunk 2]
```diff
@@ -9,6 +9,7 @@ export interface ClientPreferencesMix {
 
 const ClientPreferences = (superclass: any) => class extends superclass {
     savePreferences = async (userId: string, preferences: PreferenceType[]) => {
+        this.analytics.trackAPI('api_preferences_save_post');
```
Comment (avinashlng1080) [inline app/client/rest/preferences.ts:12]: For line 12 and 27, I need confirmation confirm if we have to track this analytics.  The event on v1 was `action_posts_flag` but I renamed it here.  Please let me know your thoughts.

File: app/screens/post_options/options/save_option/save_option.tsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    isSaved: boolean;
+    postId: string;
+}
+
+const SaveOption = ({isSaved, postId}: CopyTextProps) => {
+    const serverUrl = useServerUrl();
+
+    const onHandlePress = useCallback(async () => {
+        const remoteAction = isSaved ? deleteSavedPost : savePostPreference;
+        await remoteAction(serverUrl, postId);
```
Comment (larkox) [inline app/screens/post_options/options/save_option/save_option.tsx:24]: (Probably out of the scope of this PR) This probably is fast, but what if it is not (bad connection to the server)? The bottom sheet just stays there without any feedback (and potentially you hit the button a few more times) until the message gets through?
Shouldn't we put some "loading state"? (maybe only if it is taking too long)

**Review by larkox — APPROVED** [review]
LGTM, just a question aimed more to UX.

File: app/actions/remote/preference.ts (hunk 1 above)
Comment (enahum) [reply app/actions/remote/preference.ts:55]: @shazm can you create a ticket and handle this?

File: app/client/rest/preferences.ts (hunk 2 above)
Comment (enahum) [reply app/client/rest/preferences.ts:12]: keep the same name as it was..

File: app/screens/post_options/options/save_option/save_option.tsx (hunk 3 above)
Comment (enahum) [reply app/screens/post_options/options/save_option/save_option.tsx:24]: maybe we just remove the await on the action..

**Review by enahum — CHANGES_REQUESTED** [review]

File: app/screens/post_options/options/save_option/index.tsx [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+const enhanced = withObservables([], ({database}: WithDatabaseArgs) => {
+    const currentUserId = database.get<SystemModel>(MM_TABLES.SERVER.SYSTEM).findAndObserve(SYSTEM_IDENTIFIERS.CURRENT_USER_ID);
+
+    return {
+        currentUserId,
+    };
+});
+
+export default withDatabase(enhanced(SaveOption));
```
[the remark this answers is not in the document]
Comment (avinashlng1080) [reply app/screens/post_options/options/save_option/index.tsx]: Removed

File: app/screens/post_options/options/save_option/save_option.tsx (hunk 3 above)
Comment (avinashlng1080) [reply app/screens/post_options/options/save_option/save_option.tsx]: done

File: app/client/rest/preferences.ts (hunk 2 above)
Comment (avinashlng1080) [reply app/client/rest/preferences.ts]: done

**Review by enahum — APPROVED** [review]

**Pull request merged**